src/components/shared/SourceView-codemirror.js

```diff
@@ -0,0 +1,290 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+// @flow
+
+/**
+ * This module wraps all the interaction with the CodeMirror API into a
+ * SourceViewEditor class.
+ *
+ * This module is intended to be imported asynchronously, so that all the
+ * CodeMirror code can be split into a separate bundle chunk.
+ *
+ * This file implements the following features:
+ *  - Display source code with syntax highlighting.
+ *  - Display a gutter with:
+ *    - "Total" timings for each line
+ *    - "Self" timings for each line
+ *    - The line number for each line
+ *  - Highlight source code lines which have a non-zero timing, by applying
+ *    a cm-nonZeroLine class to them. This highlight line goes across the entire
+ *    width of the editor, it covers both the gutter and the main area.
+ */
+import { EditorView, Decoration } from '@codemirror/view';
+import {
+  EditorState,
+  StateField,
+  StateEffect,
+  Compartment,
+} from '@codemirror/state';
+import {
+  lineNumbers,
+  GutterMarker,
+  gutter,
+  gutterLineClass,
+} from '@codemirror/gutter';
+import { RangeSet } from '@codemirror/rangeset';
+import { classHighlightStyle } from '@codemirror/highlight';
+import { cpp } from '@codemirror/lang-cpp';
+import { rust } from '@codemirror/lang-rust';
+import { javascript } from '@codemirror/lang-javascript';
+
+import type { LineTimings } from 'firefox-profiler/types';
+import { emptyLineTimings } from 'firefox-profiler/profile-logic/line-timings';
+
+// This "compartment" allows us to swap the syntax highlighting language when
+// the file path changes.
+const languageConf = new Compartment();
+
+// Detect the right language based on the file extension.
+function _languageExtForPath(
+  path: string | null
+): any /* LanguageSupport | [] */ {
+  if (path === null) {
+    return [];
+  }
+  if (path.endsWith('.rs')) {
+    return rust();
+  }
+  if (path.endsWith('.js')) {
+    return javascript();
+  }
+  if (
+    path.endsWith('.c') ||
+    path.endsWith('.cc') ||
+    path.endsWith('.cpp') ||
+    path.endsWith('.cxx') ||
+    path.endsWith('.h') ||
+    path.endsWith('.hpp') ||
+    path.endsWith('.m') ||
+    path.endsWith('.mm')
+  ) {
+    return cpp();
+  }
+  return [];
+}
+
+// This gutter marker applies the "cm-nonZeroLine" class to gutter elements.
+const nonZeroLineGutterMarker = new (class extends GutterMarker {
+  elementClass = 'cm-nonZeroLine';
+})();
+
+// This "decoration" applies the "cm-nonZeroLine" class to the line of source
+// code in the main editor contents (not the gutter).
+const nonZeroLineDecoration = Decoration.line({ class: 'cm-nonZeroLine' });
+
+// An "effect" is like a redux action. This effect is
```

Comment:
Oh oops, yes there's something missing here :)

Once we switch to TypeScript, we'll also get full type coverage for this file because CodeMirror is written in TypeScript. At that point it should hopefully be even easier to understand.